### PR TITLE
JailRunner

### PR DESCRIPTION
**Pull request by youben11**
Class for running function in an environment with limited capabilities: can't import arbitrary module, only use available ones (default is torch, syft and crypten)

### CODE REVIEW

File: syft/frameworks/crypten/jail.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+        # add modules
+        for module in modules:
+            self._jail_globals[module.__name__] = module
+            self._module_names.append(module.__name__)
+
+        self._is_built = False
+        self._build()
+
+    def _build(self):
+        if self._is_built:
```
Comment (gmuraru) [inline syft/frameworks/crypten/jail.py:72]: Q: When is the ```self._is_built``` set?

File: syft/frameworks/crypten/jail.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+            raise RuntimeWarning("JailRunner already built.")
+
+        exec_globals = self._jail_globals
+        exec_globals["__builtins__"] = safe_builtins
+        exec_globals["_getiter_"] = default_guarded_getiter
+        exec_globals["_getitem_"] = default_guarded_getitem
+        exec_globals["_getattr_"] = getattr
+        # for a, b in
```
Comment (gmuraru) [inline syft/frameworks/crypten/jail.py:80]: Remove comment?

File: syft/frameworks/crypten/jail.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+        exec_globals["_iter_unpack_sequence_"] = guarded_iter_unpack_sequence
+        # for a in
```
Comment (gmuraru) [inline syft/frameworks/crypten/jail.py:82]: Also here

File: syft/frameworks/crypten/jail.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+        exec_globals["_unpack_sequence_"] = guarded_unpack_sequence
+        # unrestricted write of attr
+        exec_globals["_write_"] = lambda x: x
+        # Collecting printed strings and saved in printed local variable
+        exec_globals["_print_"] = PrintCollector
+        # exec_globals["__name__"] = "__main__"
```
Comment (gmuraru) [inline syft/frameworks/crypten/jail.py:88]: Also here :)

File: syft/frameworks/crypten/jail.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+        Returns a boolean is_func and the name of the function if is_func is True.
+        """
+
+        # The body should contain one element
+        tree = ast.parse(func_src)
+        if len(tree.body) != 1:
+            return (False, "")
+        # The body should contain a function defintion
+        func_def = tree.body[0]
+        if type(func_def) is not ast.FunctionDef:
```
Comment (gmuraru) [inline syft/frameworks/crypten/jail.py:113]: MAYBE: ```isinstance(func_def, ast.FunctionDef)```

File: test/crypten/test_jail.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
+import torch as th
+from syft.frameworks.crypten import jail
+
+
+def add_tensors(tensor):
+    # torch here is from the jail
```
Comment (gmuraru) [inline test/crypten/test_jail.py:6]: KUDOS: Nice comment :))

File: syft/frameworks/crypten/jail.py (hunk 2 above)
Comment (youben11) [reply syft/frameworks/crypten/jail.py:80]: The comments are here explaining why a certain global attribute is set

File: syft/frameworks/crypten/jail.py (hunk 1 above)
Comment (youben11) [reply syft/frameworks/crypten/jail.py:72]: Forgot to set it at the end of the build

File: pip-dep/requirements.txt [hunk 7]
```diff
@@ -6,6 +7,7 @@ numpy~=1.18.1
 phe~=1.4.0
 Pillow~=6.2.2
 requests~=2.22.0
+RestrictedPython
```
Comment (LaRiffle) [inline pip-dep/requirements.txt:10]: Add versions

File: syft/frameworks/crypten/jail.py [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+    def __init__(
+        self,
+        func_src=None,
+        func=None,
+        modules=[torch, syft, crypten],
+        rm_decorators=True,
+        **global_kwargs,
+    ):
+        """
+        Control what should be accesible from inside the function.
```
Comment (LaRiffle) [inline syft/frameworks/crypten/jail.py:29]:
```suggestion
        Control what should be accessible from inside the function.
```

File: syft/frameworks/crypten/jail.py [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+
+                    func_src = inspect.getsource(func)
+
+        if rm_decorators:
+            # Remove decorator if any
+            func_src = re.sub(r"@[^\(]+\([^\)]*\)", "", func_src)
+
+        is_func, self._func_name = JailRunner._check_func_def(func_src)
+        if not is_func:
+            raise ValueError("Not valid function definition")
```
Comment (LaRiffle) [inline syft/frameworks/crypten/jail.py:56]:
```suggestion
            raise ValueError("Not a valid function definition")
```

File: test/crypten/test_jail.py [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+    func = jail.JailRunner(func=import_socket)
+    catched = True
+
+    try:
+        func()
+        catched = False
+    except Exception as e:
+        assert isinstance(e, ImportError)
+
+    assert catched
```
Comment (LaRiffle) [inline test/crypten/test_jail.py:32]: have you checked `with pytest.raises(Error):` ?

File: pip-dep/requirements.txt (hunk 7 above)
Comment (youben11) [reply pip-dep/requirements.txt:10]: done

File: test/crypten/test_jail.py (hunk 10 above)
Comment (youben11) [reply test/crypten/test_jail.py:32]: Just learned about it, thanks for the tip

**Pull request merged**